add: NvChad 2.5 nix flake with package, overlay and home-manager module

MOIS3Y opened the pull request:
Tested on my NixOS configuration on 5 devices.

YvesStraten commented on flake.nix:
```diff
@@ -0,0 +1,50 @@
+  });
+  overlays = (import ./nix/overlays.nix { })
+  // { default = self.overlays.nvchad ; };
+  homeManagerModules.nvchad = import ./nix/module.nix;
```
please change this to/add `homeManagerModules.default` to the flake as it seems to be the norm for HM modules to have a `default` attribute

MOIS3Y replied:
@YvesStraten thanks for the review.
Are you sure the default attribute is really needed?
I have never written public home manager modules)
I based `flake.nix` on what popular modules do.
[sops-nix](https://github.com/Mic92/sops-nix/blob/master/flake.nix) does not have `.defaul`t attr
but [nix-colors](https://github.com/Misterio77/nix-colors/blob/main/default.nix)  has it listed
Looks like you're right.

I'll fix it soon.

YvesStraten replied:
> @YvesStraten thanks for the review. Are you sure the default attribute is really needed? I have never written public home manager modules) I based `flake.nix` on what popular modules do. [sops-nix](https://github.com/Mic92/sops-nix/blob/master/flake.nix) does not have `.defaul`t attr but [nix-colors](https://github.com/Misterio77/nix-colors/blob/main/default.nix) has it listed Looks like you're right.
> 
> I'll fix it soon.

Yeah, it seems to be more of personal preference, but we also make sure, if I am not mistaken, that when the user imports the module in hm, they import the `default` attr instead of having to specify it, similar to importing a `default.nix`

MOIS3Y replied:
@YvesStraten done, I pushed the changes `git push -f`
I probably should have asked first how best to submit edits, sorry)
If it is more convenient to analyze commits, I will send a new commit for all corrections

YvesStraten reviewed (changes requested):
Very good work @MOIS3Y! Everything looks good for me, I also tested it on my `aarch64-darwin` machine and it works. I am asking @siduck about whether the binary being called `nvchad` is intended. Except that, please check the comment I left, after that, this PR can be merged

YvesStraten commented on nix/nvchad.nix:
```diff
@@ -0,0 +1,68 @@
+    mkdir -p $out/{bin,config}
+    cp -r $src/* $out/config
+    install -Dm755 $nvChadBin $out/bin/nvchad
+    wrapProgram $out/bin/nvchad --prefix PATH : '${makeBinPath nativeBuildInputs}'
```
Please the binary name from `nvchad` to `nvim`, as people new or experienced with neovim will get confused when they try to use NvChad by typing `nvim` and it says not found

MOIS3Y replied:
Done

YvesStraten commented:
@siduck just pointed me to the explanation about the binary name issue, and I came up with an idea, what if we assert that `programs.neovim.enable` is `false`? That way, we don't have to worry about nvim not finding `extraPackages`, as the user will only be able to use either `programs.neovim` or `programs.nvchad`, that way, we should be able to also name the binary to `nvim` as expected

YvesStraten commented:
> > @siduck just pointed me to the explanation about the binary name issue, and I came up with an idea, what if we assert that `programs.neovim.enable` is `false` by using `lib.asserts.assertMsg`?
> > That way, we don't have to worry about nvim not finding `extraPackages`, as the user will only be able to use either `programs.neovim` or `programs.nvchad`, as a result, we should be able to also name the binary to `nvim` as expected.
> > Furthermore, we wouldn't have to use any `lib.mkForce` under the hood as suggested by the comment.
> 
> @YvesStraten `lib.asserts.assertMsg` good idea, but I can’t figure out where to place it in `module.nix` In addition, we can't prohibit the user:
> 
This can be placed in the `config` section of the module, allow me to fetch an example: [https://github.com/nix-community/home-manager/blob/465ea1f994a4e2b498b847c35caa205af7b261df/modules/services/mbsync.nix#L77][mbsync example]. As it seems, we can just add an assertions list attribute inside the `mkIf cfg.enable`, something like the following:

```nix
assertions = [
(lib.assertMsg programs.neovim.enable == false "NvChad provides a neovim binary, please choose which you want to use")
];
```

About prohibiting the user, some attributes actually do this by themselves and/or use`assertions` to prevent these conflicts, e.g. `home.sessionVariables` will prevent a variable to be 2 different values for obvious reasons, and `lib.hm.assertions.assertPlatform` will prevent a user from using a module if its architecture/platform does not match

> ```nix
> # home.nix
> home.packages = [ pkgs.neovim ];
> # configuration.nix
> environment.systemPackages = [ pkgs.neovim ];
> users.users.stepan.packages =  [ pkgs.neovim ];
> ```
> 
For `home.packages` and `environment.systemPackages` there shouldn't be a conflict as one of the binaries will go to the system binary folder, and the other will go in `/etc/profiles/per-user/<user>/bin`. Furthermore, it is not likely that a user will use both, but if a user does this, nix will prevent the generation from building anyway with an appropriate message that the binaries are conflicting.

Regarding the other solutions, I agree, they are not as good. Hope that the example I sent helps you, so that we can solve this problem.

MOIS3Y commented:
@YvesStraten thank you so much for your great advice.
I found a rather [elegant solution](https://nlewo.github.io/nixos-manual-sphinx/development/assertions.xml.html) for assertion

Everything works great
```nix
# -- Snip
  config = with pkgs; with lib; let
    confDir = "${config.xdg.configHome}/nvim";
  in mkIf cfg.enable {
    assertions = [
      {
        assertion = !config.programs.neovim.enable;
        message = ''
          NvChad provides a neovim binary, please choose which you want to use.
          
          Use Default neovim binary:
          programs.neovim.enable = true;

          Use Nvchad neovim binary:
          programs.nvchad.enable = true;

          You cannot use both at the same time.
        '';
      }
    ];
# -- Snip
```

If the user enables the standard module
```nix
  programs.neovim.enable = true;
```
He will receive the expected error and message:
```console
       error:
       Failed assertions:
       - stepan profile: NvChad provides a neovim binary, please choose which you want to use.

       Use Default neovim binary:
       programs.neovim.enable = true;

       Use Nvchad neovim binary:
       programs.nvchad.enable = true;

       You cannot use both at the same time.
```

Nice)

Maybe the text should be corrected?

I also fixed the binary, finally it's nvim 🥳

![2024-07-24-21-02-36](https://github.com/user-attachments/assets/245a18c8-3b24-468c-8053-4d4916c93f34)

I also checked what would happen if `home.packages = [pkgs.neovim]`
The new generation is generated normally, but will overwrite our executable in
`/etc/profiles/per-user/stepan/bin/nvim ⇒ /nix/store/qvh9x769112b5sv5a75sfbwf07zy8zjw-home-manager-path/bin/nvim`
 Runtime dependencies will not be found by nvim
But I completely agree with you, who would think of using the module and adding neovim to the list of user packages). So it's not a problem.


I haven't send changes or edited the README yet

What is the best way to push changes with a new commit or force push?

MOIS3Y commented:
@siduck I made all the required changes:

- made the README relevant
- the executable file is now called nvim (the user will run NvChad as usual)
- nvim.desktop allows you to launch NvChad from the launcher (uses the system icon)
- added protection if the user wants to use the standard neovim module
- updated input.nixpkgs to the latest version (nvim 0.10.0)
- checked on my `x86_64-linux` devices

@YvesStraten please check for `aarch64-darwin`

YvesStraten commented:
I shall do it today, will write if it works ASAP

YvesStraten commented:
I tested the following things: 
- `nix run` works 
- `darwin-rebuild` with hm-module creates the generation and switches over perfectly, furthermore, the binary is now named correctly (`nvim`)
- I tested the assertion by trying to enable `programs.neovim.enable`, giving an expected error message
```bash
Failed assertions:
       - yvess profile: NvChad provides a neovim binary, please choose which you want to use.

       Use Default neovim binary:
       programs.neovim.enable = true;

       Use Nvchad neovim binary:
       programs.nvchad.enable = true;

       You cannot use both at the same time.
```

Everything looks good on my `aarch64-darwin` machine, I think that this can be merged.

MOIS3Y commented:
Ha-ha, you won't believe it, but NvChad even works on Android (Pixel-7) with [nix-on-droid](https://github.com/nix-community/nix-on-droid)

<details>
  <summary>Screenshots</summary>
  <img src="https://github.com/user-attachments/assets/b1d348c9-346f-422e-86da-996c78eb5c24">
  <img src="https://github.com/user-attachments/assets/ae8a3c4a-c44d-417b-8474-1314c2aa2f6b">
  <img src="https://github.com/user-attachments/assets/e76cad80-d5dd-4e98-ad5f-216f6cef4db3">
</details>

Installed using HM in the same way as usual.
You just need to remember to tell termux to use nerd font
[issue](https://github.com/nix-community/nix-on-droid/issues/120)
```nix
# -- Snip
  terminal = {
    font = let
      jetbrains = pkgs.nerdfonts.override {
        fonts = [ "JetBrainsMono" ];
      };
      fontPath = "share/fonts/truetype/NerdFonts/JetBrainsMonoNerdFontMono-Regular.ttf";
    in "${jetbrains}/${fontPath}";
  };
# -- Snip
```

YvesStraten commented:
> Ha-ha, you won't believe it, but NvChad even works on Android (Pixel-7) with [nix-on-droid](https://github.com/nix-community/nix-on-droid)
> 
> <details>
>   <summary>Screenshots</summary>
>   <img src="https://github.com/user-attachments/assets/b1d348c9-346f-422e-86da-996c78eb5c24">
>   <img src="https://github.com/user-attachments/assets/ae8a3c4a-c44d-417b-8474-1314c2aa2f6b">
>   <img src="https://github.com/user-attachments/assets/e76cad80-d5dd-4e98-ad5f-216f6cef4db3">
> </details>
> 
> Installed using HM in the same way as usual.
> You just need to remember to tell termux to use nerd font
> [issue](https://github.com/nix-community/nix-on-droid/issues/120)
> ```nix
> # -- Snip
>   terminal = {
>     font = let
>       jetbrains = pkgs.nerdfonts.override {
>         fonts = [ "JetBrainsMono" ];
>       };
>       fontPath = "share/fonts/truetype/NerdFonts/JetBrainsMonoNerdFontMono-Regular.ttf";
>     in "${jetbrains}/${fontPath}";
>   };
> # -- Snip
> ```
> 
> 
> 

Wow, even better! 💯💯

redyf commented:
> Ha-ha, you won't believe it, but NvChad even works on Android (Pixel-7) with [nix-on-droid](https://github.com/nix-community/nix-on-droid)
> 
> <details>
>   <summary>Screenshots</summary>
>   <img src="https://github.com/user-attachments/assets/b1d348c9-346f-422e-86da-996c78eb5c24">
>   <img src="https://github.com/user-attachments/assets/ae8a3c4a-c44d-417b-8474-1314c2aa2f6b">
>   <img src="https://github.com/user-attachments/assets/e76cad80-d5dd-4e98-ad5f-216f6cef4db3">
> </details>
> 
> Installed using HM in the same way as usual.
> You just need to remember to tell termux to use nerd font
> [issue](https://github.com/nix-community/nix-on-droid/issues/120)
> ```nix
> # -- Snip
>   terminal = {
>     font = let
>       jetbrains = pkgs.nerdfonts.override {
>         fonts = [ "JetBrainsMono" ];
>       };
>       fontPath = "share/fonts/truetype/NerdFonts/JetBrainsMonoNerdFontMono-Regular.ttf";
>     in "${jetbrains}/${fontPath}";
>   };
> # -- Snip
> ```
> 
> 
> 

Now we can flex on mobile too 🗿🤝❄️

siduck commented:
is this good to merge?

YvesStraten commented:
In my opinion it is, it has been tested as well. Should be good to merge
